### PR TITLE
Add imx-dsp recipes and enable imx-gst1.0-plugin to support i.MX Audio DSP Codec

### PR DESCRIPTION
**Pull request by nsdrude**
As per NXP BSP:
- https://source.codeaurora.org/external/imx/meta-imx/tree/meta-bsp/recipes-multimedia/imx-dsp?h=hardknott-5.10.72-2.2.0
- https://source.codeaurora.org/external/imx/meta-imx/commit/?h=hardknott-5.10.72-2.2.0&id=eb72f6dc37079f001106711cdeecdb00f2d8e6cf

### CODE REVIEW

File: recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+LIC_FILES_CHKSUM = "file://COPYING;md5=03bcadc8dc0a788f66ca9e2b89f56c6f"
+
+inherit fsl-eula-unpack autotools pkgconfig
+
+SRC_URI = "${FSL_MIRROR}/${BP}.bin;fsl-eula=true"
+
+SRC_URI[md5sum] = "238d3f0256573ca657228d7090bcb7d3"
+SRC_URI[sha256sum] = "13f67f267d6d33e8be2a6c258a46cde3667258ac86435776cbf1a370de611476"
+
+EXTRA_OECONF = "-datadir=${base_libdir}/firmware --bindir=/unit_tests ${@bb.utils.contains('TUNE_FEATURES', 'aarch64', '--enable-armv8', ' ', d)}"
```
Comment (otavio) [inline recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb:14]:
```
EXTRA_OECONF += " \
    -datadir=${base_libdir}/firmware --bindir=/unit_tests \
    ${@bb.utils.contains('TUNE_FEATURES', 'aarch64', '--enable-armv8', ' ', d)} \
"
```

File: recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+HIFI4_BIN:mx8qm = "hifi4_imx8qmqxp.bin"
+HIFI4_BIN:mx8qxp = "hifi4_imx8qmqxp.bin"
+HIFI4_BIN:mx8mp = "hifi4_imx8mp.bin"
+HIFI4_BIN:mx8ulp = "hifi4_imx8ulp.bin"
+
+# No need to do install about fsl_unia.h & fsl_types.h, which are duplicate with the ones' in imx-codec
+do_install:append () {
+    if [ -d ${D}/usr/include/imx-mm/audio-codec ]; then
+        rm -rf ${D}/usr/include/imx-mm/audio-codec
+    fi
```
Comment (otavio) [inline recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb:34]: Please move the comment close to the `rm` command; also, please avoid checking if it exists and use `rm -r` only so if it stops of being installed we will know.

File: recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    # Install different DSP Firmware according to the platform
+    if [ -f ${D}/lib/firmware/imx/dsp/${HIFI4_BIN} ]; then
```
Comment (otavio) [inline recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb:36]: In which cases the file doesn't exists?

**Review by otavio — CHANGES_REQUESTED** [review]
@thochstein please review as well.

File: recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb (hunk 2 above)
Comment (nsdrude) [reply recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb:34]: @otavio I copied this recipe from [meta-imx](https://source.codeaurora.org/external/imx/meta-imx/tree/meta-bsp/recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb?h=hardknott-5.10.72-2.2.0)

As of the latest revision, `${D}/usr/include/imx-mm/audio-codec` is not being installed:
```
# After disabling do_install:append and rebuilding:
$ find tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/imx-mm
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/imx-mm/audio-codec
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/imx-mm/audio-codec/wrap
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/imx-mm/audio-codec/wrap/lib_dsp_wrap_arm_elinux.so.1
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/imx-mm/audio-codec/wrap/lib_dsp_wrap_arm_elinux.so
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/pkgconfig
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/pkgconfig/libfsldspframework.pc
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/lib
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/lib/firmware
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/lib/firmware/imx
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/lib/firmware/imx/dsp
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/lib/firmware/imx/dsp/hifi4_imx8qmqxp.bin
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/lib/firmware/imx/dsp/hifi4_imx8mp.bin
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/lib/firmware/imx/dsp/hifi4_imx8ulp.bin
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/unit_tests
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/unit_tests/DSP
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/unit_tests/DSP/dsp_test.out
```
It is not clear if this is left over from an earlier revision, or in preparation for a future revision. Do you prefer we keep the logic, or remove it for now? Changing to `rm -r ${D}/usr/include/imx-mm/audio-codec` will error right now.

File: recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb (hunk 3 above)
Comment (nsdrude) [reply recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb:36]: @otavio I copied this recipe from [meta-imx](https://source.codeaurora.org/external/imx/meta-imx/tree/meta-bsp/recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb?h=hardknott-5.10.72-2.2.0)

As of the latest revision, `${D}/lib/firmware/imx/dsp/${HIFI4_BIN}` always exists for `mx8qm|mx8qxp|mx8mp|mx8ulp`. Additionally, `hifi4_imx8mp_lpa.bin` does not exist, but it is referred to in the NXP comments.
```
# After disabling do_install:append and rebuilding:
$ find tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/imx-mm
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/imx-mm/audio-codec
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/imx-mm/audio-codec/wrap
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/imx-mm/audio-codec/wrap/lib_dsp_wrap_arm_elinux.so.1
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/imx-mm/audio-codec/wrap/lib_dsp_wrap_arm_elinux.so
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/pkgconfig
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/pkgconfig/libfsldspframework.pc
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/lib
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/lib/firmware
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/lib/firmware/imx
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/lib/firmware/imx/dsp
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/lib/firmware/imx/dsp/hifi4_imx8qmqxp.bin
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/lib/firmware/imx/dsp/hifi4_imx8mp.bin
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/lib/firmware/imx/dsp/hifi4_imx8ulp.bin
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/unit_tests
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/unit_tests/DSP
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/unit_tests/DSP/dsp_test.out
```
As with the previous comment, is not clear if this is left over from an earlier revision by NXP. I believe we can simplify to:
```
do_install:append () {
    # Rename DSP Firmware into hifi4.bin and remove unneeded binary
    mv ${D}/lib/firmware/imx/dsp/${HIFI4_BIN} ${D}/lib/firmware/imx/dsp/hifi4.bin
    find ${D}/lib/firmware/imx/dsp -name hifi4_*.bin -exec rm {} \;
}
```
Which will install for the imx8mp:
```
$ find tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/imx-mm
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/imx-mm/audio-codec
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/imx-mm/audio-codec/wrap
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/imx-mm/audio-codec/wrap/lib_dsp_wrap_arm_elinux.so.1
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/imx-mm/audio-codec/wrap/lib_dsp_wrap_arm_elinux.so
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/pkgconfig
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/usr/lib/pkgconfig/libfsldspframework.pc
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/lib
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/lib/firmware
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/lib/firmware/imx
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/lib/firmware/imx/dsp
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/lib/firmware/imx/dsp/hifi4.bin
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/unit_tests
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/unit_tests/DSP
tmp/work/imx8mp_lpddr4_evk-fslc-linux/imx-dsp/1.2.0-r0/image/unit_tests/DSP/dsp_test.out
```

File: recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb (hunk 2 above)
Comment (otavio) [reply recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb:34]: Please remove; this is why we should review.

**Comment by nsdrude** [comment]
@otavio Thanks for reviewing, I updated imx-dsp per our discussion.

File: recipes-multimedia/gstreamer/imx-gst1.0-plugin_4.6.1.bb [hunk 4]
```diff
@@ -49,6 +49,10 @@ PACKAGES =+ "${PN}-gplay ${PN}-libgplaycore ${PN}-libgstfsl ${PN}-grecorder ${PN
 # Add codec list that the beep plugin run-time depended
 BEEP_RDEPENDS = "imx-codec-aac imx-codec-mp3 imx-codec-oggvorbis"
 RDEPENDS:${PN} += "imx-parser ${BEEP_RDEPENDS} gstreamer1.0-plugins-good-id3demux "
+RDEPENDS:${PN}:append:mx8qm = " imx-dsp"
+RDEPENDS:${PN}:append:mx8qxp = " imx-dsp"
+RDEPENDS:${PN}:append:mx8mp = " imx-dsp"
+RDEPENDS:${PN}:append:mx8ulp = " imx-dsp"
```
[the remark this answers is not in the document]
Comment (nsdrude) [reply recipes-multimedia/gstreamer/imx-gst1.0-plugin_4.6.1.bb:56]: > @MaxKrummenacher I wouldn't know if on a SoC without a DSP the gstreamer plugin would even fail if it detects the dsp libraries and tries to use it.

imx-gst1.0-plugin first attempts to use the dsp wrapper: https://source.codeaurora.org/external/imx/imx-gst1.0-plugin/tree/plugins/beepdec/beepdec.c?h=MM_04.06.03_2110_L5.10.y#n439

If it fails, it attempts to use the sw wrapper interface: https://source.codeaurora.org/external/imx/imx-gst1.0-plugin/tree/plugins/beepdec/beepdec.c?h=MM_04.06.03_2110_L5.10.y#n503

Not having a DSP can be simulated by disabling the DSP remote proc framework in the device tree. With the DSP disabled, but with the libraries installed, ims-gst1.0-plulgin will try the DSP and then fallback as shown in the links above.
```
# gplay-1.0 test.mp3

FSL_GPLAY2_01.00_LINUX build on May 11 2021 03:19:55


====== BEEP: 4.6.1 build on May 11 2021 03:19:55. ======
        Core: DSP decoder Wrapper  build on Nov 24 2021 09:04:36
 file: /usr/lib/imx-mm/audio-codec/wrap/lib_dsp_wrap_arm_elinux.so

====== BEEP: 4.6.1 build on May 11 2021 03:19:55. ======
        Core: MP3 decoder Wrapper  build on Jan 11 2018 10:20:25
 file: /usr/lib/imx-mm/audio-codec/wrap/lib_mp3d_wrap_arm_elinux.so.3
CODEC: BLN_MAD-MMCODECS_MP3D_ARM_02.13.01_ARMV8  build on Aug 27 2020 17:07:39.
=========== fsl_player_play() ==================
...
```

[the remark this answers is not in the document]
Comment (MaxKrummenacher) [reply recipes-multimedia/gstreamer/imx-gst1.0-plugin_4.6.1.bb:56]: :+1:  Perfect, thank you for the explanation.

File: recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    ${@bb.utils.contains('TUNE_FEATURES', 'aarch64', '--enable-armv8', ' ', d)} \
+"
+
+RDEPENDS:${PN} += " imx-dsp-codec-ext"
+
+FILES:${PN} = "${libdir}/imx-mm/audio-codec/dsp \
+               ${libdir}/imx-mm/audio-codec/wrap \
+               ${base_libdir}/firmware/imx/dsp \
+               /unit_tests \
+"
```
Comment (otavio) [inline recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb:25]: Those lines ought to go bellow `do_install:append`.

File: recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+HIFI4_BIN ?= "hifi4_imx8qmqxp.bin"
+HIFI4_BIN:mx8qm = "hifi4_imx8qmqxp.bin"
+HIFI4_BIN:mx8qxp = "hifi4_imx8qmqxp.bin"
```
Comment (otavio) [inline recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb:29]: Those three could use the default and avoid repeating `hifi4_imx8qmqxp.bin`.

File: recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+HIFI4_BIN:mx8mp = "hifi4_imx8mp.bin"
+HIFI4_BIN:mx8ulp = "hifi4_imx8ulp.bin"
+
+do_install:append () {
+    # Rename DSP Firmware into hifi4.bin and remove unneeded binary
+    mv ${D}/lib/firmware/imx/dsp/${HIFI4_BIN} ${D}/lib/firmware/imx/dsp/hifi4.bin
+    find ${D}/lib/firmware/imx/dsp -name hifi4_*.bin -exec rm {} \;
```
Comment (otavio) [inline recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb:36]: Is this find needed?

**Review by otavio — CHANGES_REQUESTED** [review]

File: recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb (hunk 7 above)
Comment (otavio) [reply recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb:36]: I found the answer above :)

File: recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb (hunk 6 above)
Comment (nsdrude) [reply recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb]: @otavio I pushed updates to remove the repeats.

File: recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb (hunk 5 above)
Comment (nsdrude) [reply recipes-multimedia/imx-dsp/imx-dsp_1.2.0.bb]: @otavio I pushed updates moving `FILES:${PN}` below `do_install:append`.

**Review by otavio — APPROVED** [review]

**Pull request merged**